common/k8s/src/metrics_stats_aggregator.rs

```diff
@@ -0,0 +1,728 @@
+        &mut node_container_counts_map,
+    );
+    let pods_strings = format_pod_str(extended_pod_stats, controller_map);
+
+    build_node_metric_map(node_metrics, &mut node_usage_map);
+    process_nodes(
+        nodes,
+        node_usage_map,
+        &mut node_stats,
+        &mut node_pod_counts_map,
+        &mut node_container_counts_map,
+    );
+    let node_strings = format_node_str(&node_stats);
+
+    let cluster_stats = build_cluster_stats(&node_stats);
+    let cluster_stats_string = format_cluster_str(&cluster_stats);
+
+    Ok((pods_strings, node_strings, cluster_stats_string))
+}
+
+fn build_pod_metric_map(
+    pod_metrics: ObjectList<DynamicObject>,
+    pod_usage_map: &mut HashMap<String, Value>,
+) {
+    for pod_metric in pod_metrics {
+        if let Some(containers) = pod_metric.data["containers"].as_array() {
+            for container in containers {
+                let container_name = container["name"].as_str();
+
+                if container_name.is_none() {
+                    continue;
+                }
+
+                pod_usage_map.insert(
+                    container_name.unwrap().to_string(),
+                    container["usage"].clone(),
+                );
+            }
+        }
+    }
+}
+
+fn build_node_metric_map(
+    node_metrics: ObjectList<DynamicObject>,
+    node_usage_map: &mut HashMap<String, Value>,
+) {
+    for node_metric in node_metrics {
+        let node_name = node_metric
+            .metadata
+            .name
+            .unwrap_or_else(|| "NONE".to_string());
+        let usage = &node_metric.data["usage"];
+
+        node_usage_map.insert(node_name, usage.clone());
+    }
+}
+
+fn build_cluster_stats(node_stats: &Vec<NodeStats>) -> ClusterStats {
+    let mut cluster_stats = ClusterStats::new();
+
+    let mut cpu_usage = 0;
+    let mut cpu_usage_is_present = false;
+    let mut memory_usage = 0;
+    let mut memory_usage_is_present = false;
+    let mut cpu_allocatable = 0;
+    let mut cpu_allocatable_is_present = false;
+    let mut cpu_capacity = 0;
+    let mut cpu_capacity_is_present = false;
+    let mut memory_allocatable = 0;
+    let mut memory_allocatable_is_present = false;
+    let mut memory_capacity = 0;
+    let mut memory_capacity_is_present = false;
+    let mut pods_allocatable = 0;
+    let mut pods_allocatable_is_present = false;
+    let mut pods_capacity = 0;
+    let mut pods_capacity_is_present = false;
+
+    for node_stat in node_stats {
+        cluster_stats.containers_init += node_stat.containers_init;
+        cluster_stats.containers_ready += node_stat.containers_ready;
+        cluster_stats.containers_running += node_stat.containers_running;
+        cluster_stats.containers_terminated += node_stat.containers_terminated;
+        cluster_stats.containers_total += node_stat.containers_total;
+        cluster_stats.containers_waiting += node_stat.containers_waiting;
+
+        if node_stat.cpu_usage.is_some() {
```

Comment:
This is fairly verbose and repetitive, we could get rid of all of the temporaries and if statements with something like

```rust
        cluster_stats.cpu_usage = cluster_stats.cpu_usage.map_or(node_stat.cpu_usage, |current| {
            node_stat.cpu_usage.map(|new| current + new)
        });

```

That way we wouldn't need the bools or the temporary variables, we'd just accumulate straight into the cluster_stats struct

Reply:
And if we wanted to macro _that_ away we could rewrite the wholle thing to the following

```rust

macro_rules! aggregate_stat {
    ($acc_name:ident, $var_name:ident, $field_name:ident) => {
        $acc_name.$field_name = $acc_name.$field_name.map_or($var_name.$field_name, |current| {
            $var_name.$field_name.map(|new| current + new)
        });


    }
}

fn build_cluster_stats(node_stats: &Vec<NodeStats>) -> ClusterStats {
    let mut cluster_stats = ClusterStats::new();

    for node_stat in node_stats {
        cluster_stats.containers_init += node_stat.containers_init;
        cluster_stats.containers_ready += node_stat.containers_ready;
        cluster_stats.containers_running += node_stat.containers_running;
        cluster_stats.containers_terminated += node_stat.containers_terminated;
        cluster_stats.containers_total += node_stat.containers_total;
        cluster_stats.containers_waiting += node_stat.containers_waiting;

        aggregate_stat!(cluster_stats, node_stat, cpu_usage);
        aggregate_stat!(cluster_stats, node_stat, memory_usage);
        aggregate_stat!(cluster_stats, node_stat, cpu_allocatable);
        aggregate_stat!(cluster_stats, node_stat, cpu_capacity);
        aggregate_stat!(cluster_stats, node_stat, memory_allocatable);
        aggregate_stat!(cluster_stats, node_stat, memory_capacity);
        aggregate_stat!(cluster_stats, node_stat, pods_allocatable);
        aggregate_stat!(cluster_stats, node_stat, pods_capacity);

        cluster_stats.pods_failed += node_stat.pods_failed;
        cluster_stats.pods_pending += node_stat.pods_pending;
        cluster_stats.pods_running += node_stat.pods_running;
        cluster_stats.pods_succeeded += node_stat.pods_succeeded;
        cluster_stats.pods_total += node_stat.pods_total;
        cluster_stats.pods_unknown += node_stat.pods_unknown;
        cluster_stats.nodes_total += 1;

        if node_stat.ready.unwrap_or(false) {
            cluster_stats.nodes_ready += 1;
        } else {
            cluster_stats.nodes_notready += 1;
        }

        if node_stat.unschedulable.unwrap_or(false) {
            cluster_stats.nodes_unschedulable += 1;
        }
    }
    cluster_stats
}

```